assets/cases/network/NetworkWebsocket.ts

```diff
@@ -0,0 +1,195 @@
+        this._wsiSendBinary = new WebSocket('ws://localhost:8080', []);
+        this._wsiSendBinary.binaryType = 'arraybuffer';
+        this._wsiSendBinary.onopen = function (evt) {
+            respLabel.string = 'Opened!';
+            websocketLabel.string = 'WebSocket: onopen'
+        };
+
+        this._wsiSendBinary.onmessage = function (evt) {
+            const binary = new Uint8Array(evt.data);
+            let binaryStr = 'response bin msg: ';
+
+            let str = '0x';
+            const hexMap = '0123456789ABCDEF'.split('');
+            assert(hexMap.length == 16);
+
+
+            for (let i = 0; i < binary.length; i++) {
+               str += hexMap[binary[i] >> 4];
+               str += hexMap[binary[i] & 0x0F];
+            }
+
+            binaryStr += str;
+            respLabel.string = binaryStr;
+            websocketLabel.string = 'WebSocket: onmessage'
+        };
+
+        this._wsiSendBinary.onerror = function (evt) {
+           websocketLabel.string = 'WebSocket: onerror'
+            respLabel.string = 'Error!';
+        };
+         
+        this._wsiSendBinary.onclose = function (evt) {
+            websocketLabel.string = 'WebSocket: onclose'
+            // After close, it's no longer possible to use it again,
+            // if you want to send another request, you need to create a new websocket instance
+            self._wsiSendBinary = null;
+            respLabel.string = 'client is closed!';
+        };
+
+        this.scheduleOnce(this.sendWebSocketBinary, 1);
+    }
+
+    sendWebSocketBinary () {
+        let websocketLabel = this.wsStatus.node.getParent()!.getComponent(Label)!;
+        if (!this._wsiSendBinary) { return; }
+        if (this._wsiSendBinary.readyState === WebSocket.OPEN){
+            websocketLabel.string = 'WebSocket: send binary';
+            let buf = 'Hello WebSocket中文,\0 I\'m\0 a\0 binary\0 message\0.';
+
+            let arrData = new Uint16Array(buf.length);
+            for (let i = 0; i < buf.length; i++) {
+                arrData[i] = buf.charCodeAt(i);
+            }
+
+            this._wsiSendBinary.send(arrData.buffer);
+        }
+        else{
+            let warningStr = 'send binary websocket instance wasn\'t ready...';
+            websocketLabel.string = 'WebSocket: not ready';
+            this.wsStatus.string = warningStr;
+            this.scheduleOnce(()=> {
+                this.sendWebSocketBinary();
+            }, 1);
+        }
+    }
+
+    prepareWebSocketServer() {
+            const self = this;
+            const wsServerLabel = this.wsServer;
+            const respLabel = this.wsServerStatus;
+            this._wsServer = new WebSocketServer();
+
+            this._wsServer.listen(8080, (err) => {
+                 if (!err) {
+                    console.log("server booted!");
+                 } else {
+                    console.log("error when listen:", err);
+                 }
```

Comment:
Should this message be shown in the scene? I suppose the tester wouldn't check the log messages.

Reply:
good point! done.